elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/AggregationDataStoreHelper.java

```diff
@@ -195,4 +215,86 @@ private void resolveMetricMap(Set<Metric> metrics) {
+     * 1. A grouped by dimension in this query
+     * 2. An aggregated metric in this query
+     *
+     * All grouped by dimensions are defined in the entity bean, so the last entity class of a filter path
+     * must match entity class of the query.
+     *
+     * @param havingClause having clause generated from this query
+     */
+    private void validateHavingClause(FilterExpression havingClause) {
+        if (havingClause instanceof FilterPredicate) {
+            Path path = ((FilterPredicate) havingClause).getPath();
+            Path.PathElement last = path.lastElement().get();
+            Class<?> cls = last.getType();
+            String field = last.getFieldName();
+
+            if (cls != schema.getEntityClass()) {
+                throw new InvalidOperationException(
+                        String.format(
+                                "Classes don't match when try filtering on %s in having clause of %s.",
+                                cls.getSimpleName(),
+                                schema.getEntityClass().getSimpleName()));
+            }
+
+            if (schema.isMetricField(field)) {
+                Metric metric = schema.getMetric(field);
+                if (!metricMap.containsKey(metric)) {
+                    throw new InvalidOperationException(
+                            String.format(
+                                    "Metric field %s must be aggregated before filtering in having clause.", field));
+                }
+            } else {
+                if (dimensionProjections.stream().noneMatch(dim -> dim.getName().equals(field))) {
+                    throw new InvalidOperationException(
+                            String.format(
+                                    "Dimension field %s must be grouped before filtering in having clause.", field));
+                }
+            }
+        } else if (havingClause instanceof AndFilterExpression) {
+            validateHavingClause(((AndFilterExpression) havingClause).getLeft());
+            validateHavingClause(((AndFilterExpression) havingClause).getRight());
+        } else if (havingClause instanceof OrFilterExpression) {
+            validateHavingClause(((OrFilterExpression) havingClause).getLeft());
+            validateHavingClause(((OrFilterExpression) havingClause).getRight());
+        } else if (havingClause instanceof NotFilterExpression) {
+            validateHavingClause(((NotFilterExpression) havingClause).getNegated());
+        }
+    }
+
+    private void validateSorting(Sorting sorting) {
+        if (sorting != null) {
+            Map<Path, Sorting.SortOrder> sortClauses =
+                    sorting.getValidSortingRules(schema.getEntityClass(), dictionary);
+            sortClauses.entrySet().stream()
+                    .forEach((entry) -> {
+                        Path path = entry.getKey();
+                        if (path.lastElement().get().getType() == schema.getEntityClass()) {
```

Comment:
I think if `path.lastElement().get().getType() != schema.getEntityClass()` then we are sorting on a field of a relationship, then we need to make sure that relationship field is in group by dimension. For example
```
www.haha.com:4443/book?sorting=publisher.age
```
In this case we need to make sure we are grouping by publisher.